[CUDA][HIP] Fix bug in guess local worksize funcs and improve local worksize guessing in HIP adapter

hdelan opened the pull request:
The guessLocalWorkSize func for cuda adapter was erroneously giving large Y or Z factors, rounding up when it should not. This ensures that the Y and Z factors always divide the global Y or Z dimension.

It also tidies up the HIP adapter version of this func.

hdelan commented:
DPC++ PR: https://github.com/intel/llvm/pull/12663

keyradical reviewed (approved):
Looks good to me! :+1:

GeorgeWeb reviewed (commented):
Non-blocker, just a suggestion.
It maybe worth to add "tidies up the HIP adapter version of this func" to the commit message, or split the changes in two commits one per each adapter. The first one for the functional the changes in the Cuda adapter and the second one for the NFC/tidying-up changes in the HIP adapter.

Otherwise, thanks for addressing this inefficiency! LGTM

hdelan commented:
> Non-blocker, just a suggestion. It maybe worth to add "tidies up the HIP adapter version of this func" to the commit message, or split the changes in two commits one per each adapter. The first one for the functional the changes in the Cuda adapter and the second one for the NFC/tidying-up changes in the HIP adapter.
> 
> Otherwise, thanks for addressing this inefficiency! LGTM

Good shout @GeorgeWeb ! Change made

hdelan commented:
Friendly ping @ldrumm have a few patches in SYCL RT that depend on this being fixed :smile_cat:

npmiller commented:
Should the HIP adapter have the same calculation as the CUDA one?

hdelan commented:
> Should the HIP adapter have the same calculation as the CUDA one?

I suppose it would be no harm to put this in HIP adapter as well

ldrumm commented on source/adapters/cuda/enqueue.cpp:
```diff
@@ -157,30 +157,36 @@ void guessLocalWorkSize(ur_device_handle_t Device, size_t *ThreadsPerBlock,
+  // Helper lambda to make sure each x, y, z dim divide the global dimension.
+  // Can optionally specify that we want the wg size to be a power of 2 in a
+  // given dimension, which is useful for the X dim for performance reasons.
+  static auto roundToHighestFactorOfGlobalSize =
```
I really don't like functions that have switches to change their behaviour such that they're essentially different functions.
I'd also prefer that unless a function is a callback or a closure, it should be hoisted to a free function that it's easier to grep, keep function bodies short, and keep structure consistent.

hdelan replied:
Good point. I have moved this to a common place that can be shared by all adapters

ldrumm commented on source/adapters/hip/enqueue.cpp:
```diff
@@ -49,23 +49,59 @@ ur_result_t enqueueEventsWait(ur_queue_handle_t, hipStream_t Stream,
+  // Helper lambda to make sure each x, y, z dim divide the global dimension.
+  // Can optionally specify that we want the wg size to be a power of 2 in a
+  // given dimension, which is useful for the X dim for performance reasons.
+  static auto roundToHighestFactorOfGlobalSize =
```
Again here. Also, since they should be identical implementations, this could be better shared if possible - we don't want to have to maintain duplicate code everywhere.

hdelan replied:
Thanks have added to common file.

ldrumm commented on source/adapters/hip/enqueue.cpp:
```diff
@@ -49,23 +49,59 @@ ur_result_t enqueueEventsWait(ur_queue_handle_t, hipStream_t Stream,
   assert(ThreadsPerBlock != nullptr);
   assert(GlobalWorkSize != nullptr);
   assert(Kernel != nullptr);
+  int MinGrid, MaxBlockSize;
```
These are far from their first use.

hdelan replied:
Thanks, have moved

ldrumm commented on source/adapters/hip/enqueue.cpp:
```diff
@@ -49,23 +49,59 @@ ur_result_t enqueueEventsWait(ur_queue_handle_t, hipStream_t Stream,
+  size_t MaxBlockDim[3];
+
+  // The below assumes a three dimensional range but this is not guaranteed by
+  // UR.
```
We assume something that's not guaranteed? Is this a FIXME?

hdelan replied:
Thanks comment added

ldrumm reviewed (changes requested):
This changes runtime behaviour for HIP, and introduces more abstract duplicate code. Thus, I don't really think it's a refactor.

I think the `hipOccupancyMaxPotentialBlockSize` change should go in as a separate fix, and that you should revisit those lambdas

hdelan commented:
> This changes runtime behaviour for HIP, and introduces more abstract duplicate code. Thus, I don't really think it's a refactor.

I have changed the PR description to show all changes that have been included in this PR. 
 
> I think the `hipOccupancyMaxPotentialBlockSize` change should go in as a separate fix, and that you should revisit those lambdas

I would argue that it's OK to keep HIP changes in this PR as well since in UR the commits are not squashed when PRs are merged. I have constructed my commits such that there are two clear independent commits. One fixing a bug for the CUDA adapter and adding some common code, and another extending the functionality of the HIP adapter. In an ideal world maybe it would be good to split these into two PRs but with the slow pace of merging in UR I would argue to keep these changes together, especially since this PR is blocking [other PRs](https://github.com/intel/llvm/pull/12715) so need to be prioritized. Let me know your thoughts.

hdelan commented:
Friendly ping @ldrumm

ldrumm commented:
Now those lambdas are hoisted, I'm fine with this

JackAKirk commented on source/adapters/cuda/enqueue.cpp:
```diff
@@ -154,33 +153,18 @@ void guessLocalWorkSize(ur_device_handle_t Device, size_t *ThreadsPerBlock,
-    --ThreadsPerBlock[0];
-  }
+  int MinGrid, MaxBlockSize;
+  UR_CHECK_ERROR(cuOccupancyMaxPotentialBlockSize(
```
What happens here if a kernel submits a very small range e.g. 100 and has no shared memory limitations and e.g. MaxThreadsPerBlock[0] is 1024: Won't this function return a value higher than 100? Will this then be passed to the kernel execution such that it uses a range of e.g. 1024 instead of 100?

hdelan replied:
No the `roundToHighestFactorOfGlobalSizeIn3d` will make sure that no dims exceed the global dims in any dimension

JackAKirk replied:
Do you understand these failures? https://github.com/intel/llvm/actions/runs/8162721434/job/22337470532
This is the only pr that seems to affect those tests from that batch: they use parallel_for with range.

hdelan replied:
There were some bugs in the original implementation but they weren't being tested. This patch fixes the bugs and I have added some tests here https://github.com/intel/llvm/pull/12715/files#diff-0ceed90561c114d8a66c1fdfe72fdb97ac3d3af519bffabc57e62514fe117de2

But it looks like the tests failing are all just timeouts to me. I don't know what could have changed to make this happen other than hardware/CI environment related changes

hdelan replied:
You were right there were bugs in this causing an infinite loop :stuck_out_tongue_winking_eye: Have fixed

hdelan commented on test/conformance/enqueue/urEnqueueKernelLaunch.cpp:
```diff
@@ -77,56 +77,105 @@ TEST_P(urEnqueueKernelLaunchTest, InvalidWorkDimension) {
-    void SetUp() override {
-        program_name = "fill_2d";
-        UUR_RETURN_ON_FATAL_FAILURE(urKernelExecutionTest::SetUp());
+#define ENQUEUE_KERNEL_LAUNCH_TEST_1D_SIZES(SIZE)                              \
```
@kbenzie let me know what you think about these added tests. Not ideal to introduce macro funcs but I think this is the most readable way to run these tests for different sizes

kbenzie replied:
GoogleTest supports parameterized tests, it turns out similarly readable and doesn't require macros like this. I'd prefer to use GoogleTest features where possible.

hdelan replied:
Aha I will do this instead

kbenzie replied:
Here's one example of use using it for [urEnqueueUSMFill](https://github.com/oneapi-src/unified-runtime/blob/main/test/conformance/enqueue/urEnqueueUSMFill.cpp), there are others dotted around too.

hdelan commented:
> How's the intel/llvm testing looking for this?

Getting some failures on AWS that I can't reproduce locally. Slow going! Will let you know when it's sorted

Commit:
Fix bug in CUDA local range calculation

A bug in the CUDA adapter was sometimes generating Y and Z ranges that did not divide the
global Y or Z dimension. This fixes that.

Also moves some helper functions into ur/ur.hpp that may be reused by other adapters

Commit:
Make the HIP adapter use complex subgroup size calculation

The HIP adapter was only finding a good sg size in the X dim. This
    changes it so that it now chooses a sg size that divides the global dim
    in X, Y and Z dimensions. It also chooses a power of 2 sg size in the X
    dim, which is the same that the CUDA adapter does. This may give some
    performance improvements.

Commit:
Add some tests for local range finding

Dispatch kernels on lots of different configurations

hdelan commented:
@kbenzie this is good to go

kbenzie commented:
> @kbenzie this is good to go

okay cool, thanks

yingcong-wu commented:
Hi there, I think there is a problem here, which is the function prototype in `source/adapters/cuda/enqueue.hpp` not updated accordingly. 

See in https://github.com/oneapi-src/unified-runtime/blob/ed1f8bf618c88eaabea6bde0f6c06fc265f3b49f/source/adapters/cuda/enqueue.hpp#L23C52-L23C70

hdelan commented:
> Hi there, I think there is a problem here, which is the function declaration of `guessLocalWorkSize` in `source/adapters/cuda/enqueue.hpp` is not updated accordingly.
> 
> See in https://github.com/oneapi-src/unified-runtime/blob/ed1f8bf618c88eaabea6bde0f6c06fc265f3b49f/source/adapters/cuda/enqueue.hpp#L23C52-L23C70

Thanks @yingcong-wu see here https://github.com/oneapi-src/unified-runtime/pull/1460
